### ISSUE TITLE
404 with some (big?) pmtiles

### ISSUE THREAD

**Issue by eddy-geek** [protomaps/go-pmtiles#6]
TL;DR I can't get tiles out of go-pmtiles v0.5 for some files which work well through pmtiles-js.

I suspect it is linked with the file having leaf directories.

| File | Size | JS - test URL | GO - test URL |
| --- | -- | -- | -- |
| [Bugianen_Cervino](https://maki.s3.fr-par.scw.cloud/Bugianen_Cuneese.pmtiles) | 0.2 GB | :ok: | :ok:  [wtracks C-go](https://opoto.github.io/wtracks/maps.html?import=eyJJVCBCdWdpYW5lbiBDZXJ2aW5vIjp7Im9wdGlvbnMiOnsibWluWm9vbSI6IjEyIiwibWF4Wm9vbSI6IjE2IiwiYXR0cmlidXRpb24iOiIifSwidXJsIjoiaHR0cHM6Ly93d3cubW9udGFnbmUudG9wL21ha2kvQnVnaWFuZW5fQ2Vydmluby97en0ve3h9L3t5fS5qcGciLCJ0eXBlIjoiYmFzZSIsIm92ZXJsYXkiOmZhbHNlfX0=)
| [Bugianen](https://maki.s3.fr-par.scw.cloud/Bugianen.pmtiles) | 3.5 GB (sorry!) | :ok: [wtracks B-js](https://opoto.github.io/wtracks/maps.html?import=eyJJVCBCdWdpYW5lbiBqcyI6eyJvcHRpb25zIjp7Im1pblpvb20iOiIxMiIsIm1heFpvb20iOiIxNiIsImF0dHJpYnV0aW9uIjoiIn0sInVybCI6Imh0dHBzOi8vbWFraS5zMy5mci1wYXIuc2N3LmNsb3VkL0J1Z2lhbmVuLnBtdGlsZXMiLCJ0eXBlIjoicG10aWxlcyIsIm92ZXJsYXkiOmZhbHNlLCJuYW1lIjoiSVQgQnVnaWFuZW4ganMifX0=) | :no_entry: [wtracks B-go](https://opoto.github.io/wtracks/maps.html?import=eyJJVCBCdWdpYW5lbiBnbyI6eyJvcHRpb25zIjp7Im1pblpvb20iOiIxMiIsIm1heFpvb20iOiIxNiIsImF0dHJpYnV0aW9uIjoiaHR0cHM6Ly90YXJ0YW1pbGxvLndvcmRwcmVzcy5jb20vIn0sInVybCI6Imh0dHBzOi8vd3d3Lm1vbnRhZ25lLnRvcC9tYWtpL0J1Z2lhbmVuL3t6fS97eH0ve3l9LmpwZyIsInR5cGUiOiJiYXNlIiwib3ZlcmxheSI6ZmFsc2UsIm5hbWUiOiJJVCBCdWdpYW5lbiBnbyJ9fQ==)
| [eslo14_walps](https://maps.s3.fr-par.scw.cloud/eslo14_walps.pmtiles) | 1.5 G | :ok: | :no_entry:

## Context

I am integrating some pmtiles maps in https://opoto.github.io/wtracks/ 

After merging pmtiles support (opoto/wtracks#21), I have working, but excruciatingly slow pmtiles-js-on-HTTP1 maps (that we discussed in protomaps/PMTiles#16). For example  **IT Bugianen** *(find it in **More...** menu or though the custom link above)*

Now I'm trying to evaluate performance on go-pmtiles on the same map, but it does not work (metadata works 404 on tiles I tried).
See the links to download/test the files above. I including a valid file to show that the issue is not how my server setup on [montagne.top](https://montagne.top).

To test on wtracks, search for a covered area for both maps eg "Aosta".
Or test with curl:
```sh
curl -i https://www.montagne.top/maki/Bugianen_Cervino/16/34100/23382.jpg
HTTP/1.1 200 OK
curl -i https://www.montagne.top/maki/Bugianen/16/34100/23382.jpg
HTTP/1.1 404 Not Found
curl -I https://www.montagne.top/tile/eslo14_walps/16/34100/23382.png
HTTP/1.1 404 Not Found
```

I can try to debug it myself at some point, but figured maybe it's a known issue with big files?

**Comment by eddy-geek** [comment]
reproduced with master
```sh
go build
./go-pmtiles serve ./
curl http://localhost:8080/Bugianen/metadata ; echo
{"bounds": "6.592,44.024,8.789,46.498", "name": "Bugianen", "type": "baselayer", "version": "1.1", "description": "https://tartamillo.wordpress.com/bugianen/", "format": "jpg"}
curl -I http://localhost:8080/Bugianen/16/34100/23382.jpg
HTTP/1.1 404 Not Found
```
Got a bit more logs:
```
2021/12/18 04:38:41 main.go:49: Serving ./ on HTTP port: 8080 with Access-Control-Allow-Origin: 
2021/12/18 04:38:47 loop.go:143: fetched Bugianen 0-512000
2021/12/18 04:39:58 loop.go:145: failed to fetch metadata 0-512000
```

**Comment by eddy-geek** [comment]
```sh
#working
pmtiles-show Bugianen_Cervino.pmtiles
spec version:  1
metadata:
bounds = 7.295,45.706,7.734,46.012
name = Bugianen 2105 Cervino
type = baselayer
version = 1.1
description = Bugianen created on 2021-05-10 13:34:26 by MOBAC
format = jpg
root entries: 8525
```
broken
```
pmtiles-show Bugianen.pmtiles                                                                                                                               ⏎ master ✱
spec version:  1
metadata:
bounds = 6.592,44.024,8.789,46.498
name = Bugianen
type = baselayer
version = 1.1
description = https://tartamillo.wordpress.com/bugianen/
format = jpg
root entries: 3
```

So "broken" has 3 leaf directories.

I don't think I see leaf-directory support in go-pmtiles?

**Comment by bdon** [comment]
Do you have this working successfully with the python pmtiles-convert? that implementation supports writing leaf directories, while the go one doesn't yet.

**Comment by bdon** [comment]
nvm, I misread that - the go-pmtiles writer library doesn't support leaf directories yet, but the reader should. let me look more into it

**Comment by eddy-geek** [comment]
start with my very last comment -- it works, modulo the regression introduced by https://github.com/protomaps/go-pmtiles/commit/00d3f475996af96ba8a303c4566f038762bd54e9

**Comment by bdon** [comment]
OK, I understand it now. You're right that the go reader code should be permissive of the root directory having no actual tile entries and only leaf directories.

However, it does seem non-optimal to organize things that way - did you create the PMTiles archive via the python converter? can you share your original MBTiles or other source to see if the algorithm can be adapted to produce a more optimized directory layout?

**Comment by eddy-geek** [comment]
> did you create the PMTiles archive via the python converter? 
yes
> can you share your original MBTiles
[here](https://maki.s3.fr-par.scw.cloud/Bugianen.mbtiles)

**Comment by eddy-geek** [comment]
fix deployed on the server, so there's real-life testing possible with the links above now, and it's indeed quite faster than pmtiles-js thx to parallelisation & caching :-)

btw this is the code in pmtiles/writer.py that you think is "not optimized" (not such a big issue IMO)
```py
            def by_parent(t):
                if t[0] >= 7:
                    level_diff = t[0] - 7
                    return (7,t[1]//(1 << level_diff),t[2]//(1 << level_diff))
                else:
                    return (0,t[1]//(1 << t[0]),t[2]//(1 << t[0]))
            ...
            ...itertools.groupby(self.tiles,key=by_parent):
```

you're hardcoding a "base_zoom = 7", which is only only optimal for a whole world map.
instead "7" could be replaced with min zoom that fits all tiles:

```python
base_zoom = 7
n_so_far = sum(n_tiles[z] for z in range(0,8)
while  n_so_far + n_tiles[base_zoom+1] < 21845:
   n_so_far += n_tiles[base_zoom+1]
   base_zoom += 1
return base_zoom
```
(assuming we keep track of n_tiles per zoom during `write_tile`)

**Comment by eddy-geek** [comment]
this is the code in [pmtiles/writer.py](https://github.com/protomaps/PMTiles/blob/master/python/pmtiles/writer.py) that you think is "not optimized" (not such a big issue IMO)
```py
            def by_parent(t):
                if t[0] >= 7:
                    level_diff = t[0] - 7
                    return (7,t[1]//(1 << level_diff),t[2]//(1 << level_diff))
                else:
                    return (0,t[1]//(1 << t[0]),t[2]//(1 << t[0]))
            ...
            ...itertools.groupby(self.tiles,key=by_parent):
```

you're hardcoding a "base_zoom = 7", which is only only optimal for a whole world map.
instead "7" could be replaced with min zoom that fits all tiles:

```python
base_zoom = 7
n_so_far = sum(n_tiles[z] for z in range(0,8)
while  n_so_far + n_tiles[base_zoom+1] < 21845:
   n_so_far += n_tiles[base_zoom+1]
   base_zoom += 1
return base_zoom
```
(assuming we keep track of n_tiles per zoom during `write_tile`)

by the way, funnny you did a fix for the corresponding issue in pmtile-convert: https://github.com/protomaps/PMTiles/commit/f439b3bf4c9a2cc11f163a510e2ff55d809a120b *"handle case where there are not root tiles, only leaf directories"*
